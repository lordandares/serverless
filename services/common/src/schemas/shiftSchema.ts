import { InferType, array, boolean, number, object, string } from 'yup'

export const shiftEventSchema = object().shape({
  area: object().shape({
    location: object().shape({
      id: string(),
    }),
  }),
  time: string().required(),
})

export const shiftMessageSchema = object().shape({
  _id: string(),
  isGlobal: boolean(),
  message: string().required(),
  messageId: number().required(),
  punchActionId: number().required(),
  response1: string().required(),
  response1Id: number(),
  response2: string().required(),
  response2Id: number(),
  responseText: string(),
  responseTime: string(),
})

export const shiftBreakSchema = object().shape({
  end: shiftEventSchema,
  start: shiftEventSchema.required(),
})

export const shiftSchema = object().shape({
  _id: string(),
  application: string().required(),
  breaks: array().of(shiftBreakSchema),
  end: shiftEventSchema.nullable(),
  location: string().nullable(),
  messages: array().of(shiftMessageSchema),
  start: shiftEventSchema.required(),
  user: string().required(),
})

export type ShiftEventSchema = InferType<typeof shiftEventSchema>
export type ShiftMessageSchema = InferType<typeof shiftMessageSchema>
export type ShiftBreakSchema = InferType<typeof shiftBreakSchema>
export type ShiftSchema = InferType<typeof shiftSchema>

export default shiftSchema
