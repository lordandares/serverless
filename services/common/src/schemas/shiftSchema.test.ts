import { ValidationError } from 'yup'
import { shiftSchema } from './shiftSchema'

test('valid start shift', () => {
  expect.assertions(1)

  const validate = shiftSchema.validate(
    {
      application: 'applicationId',
      end: null,
      location: 'locationId',
      start: {
        time: 'timestamp',
      },
      user: 'userId',
    },
    { strict: true },
  )

  return expect(validate).resolves.toMatchSnapshot()
})

test('valid end shift', () => {
  expect.assertions(1)

  const validate = shiftSchema.validate(
    {
      application: 'applicationId',
      end: {
        time: 'end-timestamp',
      },
      location: 'locationId',
      start: {
        time: 'start-timestamp',
      },
      user: 'userId',
    },
    { strict: true },
  )

  return expect(validate).resolves.toMatchSnapshot()
})

test('valid shift for GPS user', () => {
  expect.assertions(1)

  const validate = shiftSchema.validate(
    {
      application: 'applicationId',
      end: {
        area: {
          location: {
            id: 'area-location-id-2',
          },
        },
        time: 'end-timestamp',
      },
      location: null,
      start: {
        area: {
          location: {
            id: 'area-location-id-1',
          },
        },
        time: 'start-timestamp',
      },
      user: 'userId',
    },
    { strict: true },
  )

  return expect(validate).resolves.toMatchSnapshot()
})

test('valid message', () => {
  expect.assertions(1)

  const validate = shiftSchema.validate(
    {
      application: 'applicationId',
      end: null,
      location: 'locationId',
      messages: [
        {
          isGlobal: false,
          messageId: 123,
          message: 'message',
          punchActionId: 0,
          response1: 'response1',
          response1Id: 0,
          response2: 'response2',
          response2Id: 0,
        },
      ],
      start: {
        time: 'timestamp',
      },
      user: 'userId',
    },
    { strict: true },
  )

  return expect(validate).resolves.toMatchSnapshot()
})

test('invalid message', () => {
  expect.assertions(1)

  const validate = shiftSchema.validate(
    {
      application: 'applicationId',
      end: null,
      location: 'locationId',
      messages: [
        {
          id: '',
        },
      ],
      start: {
        time: 'timestamp',
      },
      user: 'userId',
    },
    { strict: true },
  )

  return expect(validate).rejects.toThrowError(ValidationError)
})

test('missing application', () => {
  expect.assertions(1)

  const validate = shiftSchema.validate(
    {
      application: '',
      end: {
        area: {
          location: {
            id: 'area-location-id-2',
          },
        },
        time: 'end-timestamp',
      },
      start: {
        area: {
          location: {
            id: 'area-location-id-1',
          },
        },
        time: 'start-timestamp',
      },
      user: 'userId',
    },
    { strict: true },
  )

  return expect(validate).rejects.toThrowError(ValidationError)
})

test('missing user', () => {
  expect.assertions(1)

  const validate = shiftSchema.validate(
    {
      application: 'applicationId',
      end: {
        area: {
          location: {
            id: 'area-location-id-2',
          },
        },
        time: 'end-timestamp',
      },
      start: {
        area: {
          location: {
            id: 'area-location-id-1',
          },
        },
        time: 'start-timestamp',
      },
      user: '',
    },
    { strict: true },
  )

  return expect(validate).rejects.toThrowError(ValidationError)
})
