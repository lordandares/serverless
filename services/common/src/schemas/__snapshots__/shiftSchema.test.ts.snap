// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`valid end shift 1`] = `
Object {
  "application": "applicationId",
  "end": Object {
    "time": "end-timestamp",
  },
  "location": "locationId",
  "start": Object {
    "time": "start-timestamp",
  },
  "user": "userId",
}
`;

exports[`valid message 1`] = `
Object {
  "application": "applicationId",
  "end": null,
  "location": "locationId",
  "messages": Array [
    Object {
      "isGlobal": false,
      "message": "message",
      "messageId": 123,
      "punchActionId": 0,
      "response1": "response1",
      "response1Id": 0,
      "response2": "response2",
      "response2Id": 0,
    },
  ],
  "start": Object {
    "time": "timestamp",
  },
  "user": "userId",
}
`;

exports[`valid shift for GPS user 1`] = `
Object {
  "application": "applicationId",
  "end": Object {
    "area": Object {
      "location": Object {
        "id": "area-location-id-2",
      },
    },
    "time": "end-timestamp",
  },
  "location": null,
  "start": Object {
    "area": Object {
      "location": Object {
        "id": "area-location-id-1",
      },
    },
    "time": "start-timestamp",
  },
  "user": "userId",
}
`;

exports[`valid start shift 1`] = `
Object {
  "application": "applicationId",
  "end": null,
  "location": "locationId",
  "start": Object {
    "time": "timestamp",
  },
  "user": "userId",
}
`;
